article/_posts/2023-01-19-DreamerV3.md

```diff
@@ -0,0 +1,113 @@
+![](/article/images/dreamerv3/cat_dist.jpeg)
+> https://ai.googleblog.com/2021/02/mastering-atari-with-discrete-world.html
+
+The RSSM is trained via a set of three losses:
+
+![](/article/images/dreamerv3/eq4.jpeg)
+
+![](/article/images/dreamerv3/eq5.jpeg)
+
+$$\mathcal{L}_{dyn}$$ and $$\mathcal{L}_{rep}$$ make sure that the two representations match and are not trivial through a process called *KL balancing*.
+
+The Actor-Critic is trained to propose actions and estimate their return solely on a the latent space predicted by the dynamics predictor. This way, only a single input state is needed to evaluate actions in the future. The actor is trained using REINFORCE and the critic is trained using $$\lambda$$-returns.
+
+To train the model, the environment is explored for some steps, which are then used to train the world-model, then the actor-critic. Then, the environment is explored some more and the process repeats. The total number of steps perfomed depends on the environment. The gradients of the actor-critic do not backpropagate through the world model.
+
+## Tricks
+
+While the architecture of DreamerV2-3 is somewhat straightforward, several tricks are employed to make the training robust to the wide variety of environments.
+
+#### Symlog
+
+Different environments may produce different reward distributions. Some may output very large and sparse rewards, others small by very dense rewards. As such, there needs to be some form of normalization to provide the agent with a similar loss amplitude across all environments. To do so, the authors introduce the symlog function (and its inverse, symexp):
+
+![](/article/images/dreamerv3/eq2.jpeg)
+
+![](/article/images/dreamerv3/fig4.jpeg)
+
+The symlog function squashes large positive and negative values while preserving small values. The symlog function is used "in the decoder, the reward predictor, and the critic. It also squashes inputs to the encoder[sic]".
+
+While there is no explicit mention as to why this is done, the symlog function is not applied directly. Instead, a network is trained to predict the symlog function such that
+
+![](/article/images/dreamerv3/eq1.jpeg)
+
+#### Free bits
+
+Readers with a keen eye may have spotted that the KL-balancing terms are clipped below 1 nat. The authors argue this is to "*avoid a degenerate solution where the dynamics are trivial to predict but contain not enough information about the inputs. [...] This disables them while they are already minimized well to focus the world model on its prediction loss*".
+
+### Two-hot reward encoding
+
+In environments with a widespread distribution of returns, predicting the expected returns by the critic might not be descriptive enough to train the actor. Instead, the rewards are transformed into two-hot encoding such that:
+
+![](/article/images/dreamerv3/eq8.jpeg)
+![](/article/images/dreamerv3/eq9_10.jpeg)
+
+Other tricks are employed but I did not think they were worth mentionning.
```

Comment:
mentionning -> mentioning

Reply:
Fixed in [9165505](https://github.com/vitalab/vitalab.github.io/pull/584/commits/9165505ec5c9dd0b82d1c86d9af558581a71ab30)